src/Clapi/Relay.hs

```diff
@@ -201,96 +134,419 @@ mayContDiff ma kb = case ma of
   where
-    fwd (i, dig) = case dig of
-        PnidRootGet -> do
-            sendRev
-              ( i
-              , Ocrid $ FrcRootDigest $ Map.fromSet (const $ SoAfter Nothing) $
-                Map.keysSet vsm)
-            relay vsm
-        PnidClientGet cgd ->
-          let (ocsed, ocids) = genInitDigests cgd vsm in do
-            unless (ocsedNull ocsed) $ sendRev (i, Ocsed ocsed)
-            mapM_ (sendRev . (i,) . Ocid) ocids
-            relay vsm
-        PnidTrcud trcud -> let ns = trcudNamespace trcud in maybe
-          (do
-            sendRev (i, Ocud $ (frcudEmpty ns)
-              -- FIXME: Not a global error, NamespaceError?
-              { frcudErrors = Map.singleton GlobalError
-                ["fixthis: Bad namespace"]
-              })
-            relay vsm)
-          (\vs -> let (ocud, opd) = handleTrcud vs trcud in do
-            sendRev (i, Opd $ opd)
-            sendRev (i, Ocud $ ocud)
-            relay vsm
-          )
-          (Map.lookup ns vsm)
-        PnidTrpd trpd ->
-          let
-            ns = trpdNamespace trpd
-            vs = Map.findWithDefault
-                (baseValuespace (Tagged $ unNamespace ns) Editable) ns vsm
-          in
-            either terminalErrors (handleOwnerSuccess trpd vs) $
-              processToRelayProviderDigest trpd vs
-        PnidTrprd (TrprDigest ns) -> do
-          sendRev (i, Ocrd $ FrcRootDigest $ Map.singleton ns SoAbsent)
-          relay $ Map.delete ns vsm
-      where
-        handleOwnerSuccess
-            (TrpDigest ns postDefs defs dd contOps errs)
-            vs (updatedTyAssns, vs') =
+    fwd (ClientConnect _ i) = handleConnect i
+    fwd (ClientData i trd) = case trd of
+      Trpd d -> handleTrpd i d
+      Trprd d -> handleTrprd i d
+      Trcsd d -> handleTrcsd i d
+      Trcud d -> handleTrcud i d
+    fwd (ClientDisconnect i) = handleDisconnect i
+
+handleConnect
+  :: (Ord i, Monad m) => i -> StateT (RelayState i) (RelayProtocol i m) ()
+handleConnect i = do
+  modify $ over rsAllClients $ Set.insert i
+  vsm <- view rsVsMap <$> get
+  sendData' i $ Frcrd $ FrcRootDigest $ Map.fromSet
+    (const $ SoAfter Nothing) $ Map.keysSet vsm
+
+handleTrpd
+  :: (Ord i, Monad m)
+  => i -> TrpDigest -> StateT (RelayState i) (RelayProtocol i m) ()
+handleTrpd i d = do
+    (existingOwner, newOwners) <-
+      Map.insertLookupWithKey (\_ v _ -> v) ns i . view rsOwners <$> get
+    result <- runExceptT $ case existingOwner of
+      Nothing -> do
+        -- NB: as long as we make sure the initial claim has data in it anywhere
+        -- we can never subequently have an empty valuespace:
+        when (null $ trpdData d) $ throwError $
+          Map.singleton (NamespaceError ns) ["Empty namespace claim"]
+        frcud <- tryVsUpdate
+        lift $ updateOwners newOwners
+        -- FIXME: would be nice if we didn't have to track ourselves that we'd
+        -- added a new Namespace to the Valuepace Map...
```

Comment:
I guess that would be making the VS map bit a `Map Namespace (i, Valuespace)`?
Not saying we should go there now.
Also Valuepace.